### PR TITLE
VB Service Added - Inline Parameter Name Hints

### PR DESCRIPTION
**Pull request by akhera99**
Hello,

This pull request is to be able to merge the added Visual Basic service to the current feature branch I have out. 

View of the VB hints:
image

Modified/Created Files:
```InlineParameterNameHintsService.vb``` much like the service for C#, I go through and located all the SimpleArgumentSyntax types and determine if the adornment should be added. 
```VisualBasicInlineParameterNameHintsTests.vb``` essentially did all the tests I did for the C# tagger, but for VB.
```ArgumentSyntaxExtensions.vb``` previously the DetermineParameter function made a call to the GetSymbolInfo method which had the type ExpressionSyntax as a parameter. I just changed the call to use the GetSymbolInfo method which had the SyntaxNode type as a parameter so that I would be able to get the associated parameter for expressions as well as attributes.

Please let me know if you have any questions or concerns!

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Extensions/ArgumentSyntaxExtensions.vb [hunk 1]
```diff
@@ -31,12 +31,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
                 Return Nothing
             End If
 
-            Dim invocableExpression = TryCast(argumentList.Parent, ExpressionSyntax)
-            If invocableExpression Is Nothing Then
-                Return Nothing
-            End If
-
-            Dim symbol = semanticModel.GetSymbolInfo(invocableExpression, cancellationToken).Symbol
+            ' Gets the symbol info for the SyntaxNode
```
Comment (jasonmalinowski) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Extensions/ArgumentSyntaxExtensions.vb:34]: You probably don't need this comment here since yes that's just what GetSymbolInfo is doing.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            Dim tree = Await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(False)
+            Dim root = Await tree.GetRootAsync(cancellationToken).ConfigureAwait(False)
+            Dim spans = New List(Of InlineParameterHint)
+
+            Dim semanticModel = Await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(False)
+            Dim nodes = root.DescendantNodes()
+
+            For Each node In nodes
+                If TypeOf node Is SimpleArgumentSyntax Then
+                    Dim simpleArgument = TryCast(node, SimpleArgumentSyntax)
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:33]: In this case, just do the TryCast first and check that for being null; that saves an extra type check.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    If Not simpleArgument.IsNamed AndAlso simpleArgument.NameColonEquals Is Nothing AndAlso IsExpressionWithNoName(simpleArgument.Expression) Then
+                        Dim param = simpleArgument.DetermineParameter(semanticModel, False, cancellationToken)
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:35]: As a deeply meta point, use a named parameter for the "False" since we can't tell what this. You can also choose to resolve this comment via adding your feature to GitHub but that might take some extra work.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        Public Async Function GetInlineParameterNameHintsAsync(document As Document, textSpan As TextSpan, cancellationToken As CancellationToken) As Task(Of IEnumerable(Of InlineParameterHint)) Implements IInlineParameterNameHintsService.GetInlineParameterNameHintsAsync
+            Dim tree = Await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(False)
+            Dim root = Await tree.GetRootAsync(cancellationToken).ConfigureAwait(False)
+            Dim spans = New List(Of InlineParameterHint)
+
+            Dim semanticModel = Await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(False)
+            Dim nodes = root.DescendantNodes()
+
+            For Each node In nodes
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:31]: Something in this loop should be checking for cancellation.

File: src/EditorFeatures/Test2/InlineParameterNameHints/VisualBasicInlineParameterNameHintsTests.vb [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+        <WpfFact, Trait(Traits.Feature, Traits.Features.InlineParameterNameHints)>
+        Public Async Function TestLiteralNestedCastParametersSimpleCase() As Task
+            Dim input =
+            <Workspace>
+                <Project Language="Visual Basic" CommonReferences="true">
+                    <Document>
+                        Class Foo
+                            Sub Main(args As String())
+                                TestMethod({|x:CInt(5.5)|}, {|y:2.2|})
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/InlineParameterNameHints/VisualBasicInlineParameterNameHintsTests.vb:105]: Also add tests for CType, TryCast and DirectCast.

File: src/EditorFeatures/Test2/InlineParameterNameHints/AbstractInlineParameterNameHintsTests.vb [hunk 6]
```diff
@@ -25,7 +25,7 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.InlineParameterNameHints
                 Dim hostDocument = workspace.Documents.Single()
                 Dim snapshot = hostDocument.GetTextBuffer().CurrentSnapshot
                 Dim document = workspace.CurrentSolution.GetDocument(hostDocument.Id)
-                Dim tagService = document.GetRequiredLanguageService(Of IInlineParameterNameHintsService)
+                Dim tagService = document.GetLanguageService(Of IInlineParameterNameHintsService)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/InlineParameterNameHints/AbstractInlineParameterNameHintsTests.vb:28]: Why the change?

Comment (akhera99) [reply src/EditorFeatures/Test2/InlineParameterNameHints/AbstractInlineParameterNameHintsTests.vb:28]: So I forgot to change it back. I had thought that was the reason my tests weren't running, but I had not changed the language type for each test itself to VB.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+Imports System.Composition
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.Host.Mef
+Imports Microsoft.CodeAnalysis.InlineParameterNameHints
+Imports Microsoft.CodeAnalysis.Text
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.InlineParameterNameHints
+
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:13]:
```suggestion
```

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    <ExportLanguageService(GetType(IInlineParameterNameHintsService), LanguageNames.VisualBasic), [Shared]>
+    Friend Class InlineParameterNameHintsService
+        Implements IInlineParameterNameHintsService
+
+        <ImportingConstructor>
+        <Obsolete(MefConstruction.ImportingConstructorMessage, True)>
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:19]: named parameter for true.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                        Dim param = simpleArgument.DetermineParameter(semanticModel, False, cancellationToken)
+                        If param IsNot Nothing AndAlso param.Name.Length > 0 Then
+                            spans.Add(New InlineParameterHint(param.Name, simpleArgument.Span.Start))
+                        End If
+                    End If
+                End If
+            Next
+
+            Return spans
+
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:44]:
```suggestion
```

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                If TypeOf node Is SimpleArgumentSyntax Then
+                    Dim simpleArgument = TryCast(node, SimpleArgumentSyntax)
+                    If Not simpleArgument.IsNamed AndAlso simpleArgument.NameColonEquals Is Nothing AndAlso IsExpressionWithNoName(simpleArgument.Expression) Then
+                        Dim param = simpleArgument.DetermineParameter(semanticModel, False, cancellationToken)
+                        If param IsNot Nothing AndAlso param.Name.Length > 0 Then
+                            spans.Add(New InlineParameterHint(param.Name, simpleArgument.Span.Start))
+                        End If
+                    End If
+                End If
+            Next
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:41]: this loop looks effectively equivalent with teh c# version.  it would be nice to have this type (and hte C# version) derive from an abstract implementation that has the core algorithm, but calls out to the subclasses to provide lang-specific work that cannot be shared.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+            If TypeOf arg Is UnaryExpressionSyntax Then
+                Dim negation = TryCast(arg, UnaryExpressionSyntax)
+                ' Recurse until we find a literal
+                ' If so, then we should add the adornment
+                Return IsExpressionWithNoName(negation.Operand)
+            End If
+
+            Return False
+
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:75]:
```suggestion
```

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            If TypeOf arg Is PredefinedCastExpressionSyntax Then
+                Dim cast = TryCast(arg, PredefinedCastExpressionSyntax)
+                ' Recurse until we find a literal
+                ' If so, then we should add the adornment
+                Return IsExpressionWithNoName(cast.Expression)
+            End If
+
+            If TypeOf arg Is UnaryExpressionSyntax Then
+                Dim negation = TryCast(arg, UnaryExpressionSyntax)
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:68]:
```suggestion
                Dim negation = directCast(arg, UnaryExpressionSyntax)
```

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+            If TypeOf arg Is ObjectCreationExpressionSyntax Then
+                ' We want to adorn object invocations that exist as arguments because they are Not declared anywhere
+                ' else in the file
+                ' Example: testMethod(^ New Object()); should show the adornment at the caret  
+                Return True
+            End If
+
+            If TypeOf arg Is PredefinedCastExpressionSyntax Then
+                Dim cast = TryCast(arg, PredefinedCastExpressionSyntax)
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:61]:
```suggestion
                Dim cast = directCast(arg, PredefinedCastExpressionSyntax)
```

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb (hunk 8 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:19]: It does not allow me to add one.

File: src/EditorFeatures/Test2/InlineParameterNameHints/VisualBasicInlineParameterNameHintsTests.vb (hunk 5 above)
Comment (akhera99) [reply src/EditorFeatures/Test2/InlineParameterNameHints/VisualBasicInlineParameterNameHintsTests.vb:105]: Added, they don't resolve to the same base type so I had to add checks for it in the service as well.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb (hunk 4 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:23]: Added

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb (hunk 10 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:34]: Please let me know if the changes I made abstracted it out enough

File: src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InlineParameterNameHints
+{
+    internal abstract class AbstractInlineParameterNameHintsService : IInlineParameterNameHintsService
+    {
+        public async Task<IEnumerable<InlineParameterHint>> GetInlineParameterNameHintsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var spans = new List<InlineParameterHint>();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs:17]: remove this and just do `var spans = ...` below.

File: src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var nodes = root.DescendantNodes(textSpan);
+            spans = AddAllParameterNameHintLocations(semanticModel, nodes, spans, cancellationToken);
+            return spans;
+        }
+
+        protected abstract List<InlineParameterHint> AddAllParameterNameHintLocations(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs:26]: don't really see why this is returning a List.  Either have it be IEnuemrable, or ImmutableArray.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs:26]: oh wait.  it's even weirder :)   you both pass in teh value to add to, and you return it out as a return value.  Pick one or the other :)

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:18]: really?  weird :(

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:34]: yup.  i would consider this good enough sharing :)

File: src/Features/CSharp/Portable/InlineParameterNameHints/InlineParameterNameHintsService.cs [hunk 16]
```diff
@@ -23,26 +23,17 @@ namespace Microsoft.CodeAnalysis.CSharp.InlineParameterNameHints
     /// as well as associate the adornments back to the parameter name
     /// </summary>
     [ExportLanguageService(typeof(IInlineParameterNameHintsService), LanguageNames.CSharp), Shared]
-    internal class InlineParameterNameHintsService : IInlineParameterNameHintsService
+    internal class InlineParameterNameHintsService : AbstractInlineParameterNameHintsService
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InlineParameterNameHints/InlineParameterNameHintsService.cs:26]: once we have a VB and C# version we rename thsee.  so this should be CSharpInlineParameterNameHintsService.  That way it's easy to find/nav-to them since we have distinct names, not three types with the same name.

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+Imports System.Composition
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.Host.Mef
+Imports Microsoft.CodeAnalysis.InlineParameterNameHints
+Imports Microsoft.CodeAnalysis.Text
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.InlineParameterNameHints
+    <ExportLanguageService(GetType(IInlineParameterNameHintsService), LanguageNames.VisualBasic), [Shared]>
+    Friend Class InlineParameterNameHintsService
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:14]: rename

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        Protected Overrides Function AddAllParameterNameHintLocations(semanticModel As SemanticModel, nodes As IEnumerable(Of SyntaxNode), spans As List(Of InlineParameterHint), cancellationToken As CancellationToken) As List(Of InlineParameterHint)
+            For Each node In nodes
+                cancellationToken.ThrowIfCancellationRequested()
+                Dim simpleArgument = TryCast(node, SimpleArgumentSyntax)
+                If Not simpleArgument Is Nothing Then
+                    If Not simpleArgument.IsNamed AndAlso simpleArgument.NameColonEquals Is Nothing AndAlso IsExpressionWithNoName(simpleArgument.Expression) Then
+                        Dim param = simpleArgument.DetermineParameter(semanticModel, allowParamArray:=False, cancellationToken)
+                        If param IsNot Nothing AndAlso param.Name.Length > 0 Then
+                            spans.Add(New InlineParameterHint(param.Name, simpleArgument.Span.Start))
+                        End If
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/InlineParameterNameHintsService.vb:31]: nit: you could share this code with c# by having a protected helper in teh base class.  up to you.

File: src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs (hunk 14 above)
Comment (akhera99) [reply src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs:17]: Fixed

File: src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs (hunk 15 above)
Comment (akhera99) [reply src/Features/Core/Portable/InlineParameterNameHints/AbstractInlineParameterNameHintsService.cs:26]: fixed

File: src/Features/VisualBasic/Portable/InlineParameterNameHints/VisualBasicInlineParameterNameHintsService.vb [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+                ' If so, then we should add the adornment
+                Return IsExpressionWithNoName(cast.Expression)
+            End If
+
+            If TypeOf arg Is DirectCastExpressionSyntax Then
+                Dim cast = DirectCast(arg, DirectCastExpressionSyntax)
+                ' Recurse until we find a literal
+                ' If so, then we should add the adornment
+                Return IsExpressionWithNoName(cast.Expression)
+            End If
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/InlineParameterNameHints/VisualBasicInlineParameterNameHintsService.vb:79]: There's a CastExpressionSyntax base type you can use to cover all of these if you wanted.
